### PR TITLE
untested fix for MixinEntity including clientlevel

### PR DESCRIPTION
**Pull request by m3t4f1v3**
# ⚒️ Pull Request Offering

> _We shall not introduce more regressions  \
> Lest we face eternal digressions_

---
## 🧾 What This PR Does
- Bug fix?  yesss
- Feature?  
- Refactor?  
- Data/config update?  

## 🔍 How This Was Tested
- [ ] GameTest framework  
- [ ] Singleplayer / Server  
- [ ] Logs looked sane  
- [x] The vibes felt right *(sometimes you really do just have to vibe it)*

## ⚠️ Breaking Changes
- [ ] Yes (describe)  
- [x] No  (are you *really* sure?)

---

## 🧪 GameTest Oath
- [ ] Please include at least one `@GameTest`  
or  
- [x] PR does not require a `@GameTest`   
  - [ ] True if this change is only a data/config/docs/logs change  
  - [x] Also True *if* you can make a good case on why you shouldn't have to add one

### 📍 GameTest Details (Optional Unless Required)
Provide the location and intent of the test(s):

- **Test class:** `com.example.mod.gametest.YourTestClass`
- **Test name(s):**
  - `test_something_does_not_break_reality`
- **What it verifies:**
  - *(Brief explanation)*

> PRs without a new GameTest **will not be merged**  (If they are determined to be required) \
> _unless explicitly exempted by a maintainer._

---

## 🗝️ Additional Notes
Anything else reviewers might need to know:
performance concerns, edge cases, forbidden knowledge, etc. 💀

---

## 🜏 Declaration
By submitting this PR, I affirm:  
- Code/data compiles and loads  
- Tests _(if any)_ pass  
- No worlds shall be harmed in the making or merging of this PR 💀

### CODE REVIEW

**Comment by blockninja124** [comment]
Can you explain what this is and why?  Also https://github.com/ValkyrienSkies/Valkyrien-Skies-2/pull/1504 exists for EMF fixes

**Comment by jacky3362** [comment]
he made this fix cause i got a crash on server env of the attached log im adding
a server crash due to trying to load client stuff on server env vs 2.4.0
[latest.log](https://github.com/user-attachments/files/24343059/latest.log)

**Comment by m3t4f1v3** [comment]
oh my bad, i'll force push it without the gradle changes, that was for a previous commit that fixed emf stuff but theirs looks better

File: common/src/main/java/org/valkyrienskies/mod/mixin/entity/MixinEntity.java [hunk 1]
```diff
@@ -211,28 +210,6 @@ private void preCalculateViewVector(final float xRot, final float yRot, final Ca
...
     */
 
-    /**
-     * @reason Without this and that other mixin, things don't render correctly at high speeds.
-     * @see org.valkyrienskies.mod.mixin.client.renderer.MixinEntityRenderer
-     */
-    @Inject(method = "shouldRender", at = @At("HEAD"), cancellable = true)
-    private void onShouldRender(double d, double e, double f, CallbackInfoReturnable<Boolean> cir) {
-        if (this.draggingInformation.isEntityBeingDraggedByAShip() && this.level.isClientSide) {
-            final ClientShip ship = VSGameUtilsKt.getShipObjectWorld((ClientLevel) this.level).getAllShips().getById(this.draggingInformation.getLastShipStoodOn());
```
Comment (zyxkad) [inline common/src/main/java/org/valkyrienskies/mod/mixin/entity/MixinEntity.java:221]: I believe you should do `VSGameUtilsKt.getShipObjectWorld(this.level).getShip... instanceof ClientShip ship` instead of create a separate mixin or use force casting. One should not assume `this.level` is always `ClientLevel` for compatible reasons.

Comment (m3t4f1v3) [reply common/src/main/java/org/valkyrienskies/mod/mixin/entity/MixinEntity.java:221]: importing ClientLevel crashes the server

Comment (zyxkad) [reply common/src/main/java/org/valkyrienskies/mod/mixin/entity/MixinEntity.java:221]: > importing ClientLevel crashes the server

so you do not import it, you just do `VSGameUtilsKt.getShipObjectWorld(this.level)` directly

File: common/src/main/java/org/valkyrienskies/mod/mixin/entity/MixinEntity.java [hunk 2]
```diff
@@ -218,9 +217,9 @@ private void preCalculateViewVector(final float xRot, final float yRot, final Ca
     @Inject(method = "shouldRender", at = @At("HEAD"), cancellable = true)
     private void onShouldRender(double d, double e, double f, CallbackInfoReturnable<Boolean> cir) {
         if (this.draggingInformation.isEntityBeingDraggedByAShip() && this.level.isClientSide) {
-            final ClientShip ship = VSGameUtilsKt.getShipObjectWorld((ClientLevel) this.level).getAllShips().getById(this.draggingInformation.getLastShipStoodOn());
-            if (ship != null) {
-                final ShipTransform shipTransform = ship.getRenderTransform();
+            final Ship ship = VSGameUtilsKt.getShipObjectWorld(this.level).getAllShips().getById(this.draggingInformation.getLastShipStoodOn());
+            if (ship != null && ship instanceof final ClientShip clientShip) {
+                final ShipTransform shipTransform = clientShip.getRenderTransform();
```
Comment (zyxkad) [inline common/src/main/java/org/valkyrienskies/mod/mixin/entity/MixinEntity.java]: I just noticed, if anything invoked `shouldRender` on server side, they'll also expect it to work correctly. So maybe do a check here:

```suggestion
            if (ship != null) {
                final ShipTransform shipTransform = ship instanceof final ClientShip clientShip ? clientShip.getRenderTransform() : ship.getTransform();
```

**Comment by m3t4f1v3** [comment]
closed by https://github.com/ValkyrienSkies/Valkyrien-Skies-2/pull/1512

**Comment by ThePlasticPotato** [comment]
Somehow didn’t notice this gg
